### ISSUE TITLE
"Received RST_STREAM with error code 2" from any call

### ISSUE THREAD

**Issue by nbellamy-paper** [googleapis/python-datastream#123]
#### Environment details

  - OS type and version: OS X 12.3.1
  - Python version: `python --version`: Python 3.9.13
  - pip version: `pip --version`: pip 22.2.2
  - `google-cloud-datastream` version: `pip show google-cloud-datastream`: 1.1.0

#### Steps to reproduce

  1. Fresh virtualenv, installing nothing but `pip install google-cloud-datastream`
  2. Use the sample code at https://github.com/googleapis/python-datastream/blob/main/samples/generated_samples/datastream_v1_generated_datastream_get_connection_profile_sync.py (or any code that actually makes a datastream API call)
  3. Receive error
```
google.api_core.exceptions.InternalServerError: 500 Received RST_STREAM with error code 2
```

I'm using the example here of DatastreamClient().get_connection_profile(), but I've received the same error from multiple other methods like list_connection_profiles(), list_routes(), anything that calls the API basically.

#### Code example
Simplest:
```python
from google.cloud import datastream_v1
client = datastream_v1.DatastreamClient()
response = client.get_connection_profile(name="mysql-bq-dev")
```

Following the sample code:
```python
from google.cloud import datastream_v1
client = datastream_v1.DatastreamClient()
request = datastream_v1.GetConnectionProfileRequest(name="mysql-bq-dev")
response = client.get_connection_profile(request=request)
```
Both yield the same error. Going line by line, declaring the client is fine, declaring the request is fine, but once client.get_connection_profile() is called the error is thrown.

#### Stack trace
From 2nd example above
```
Traceback (most recent call last):
  File "/Users/nick.bellamy/test/datastream/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 50, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/Users/nick.bellamy/test/datastream/lib/python3.9/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/Users/nick.bellamy/test/datastream/lib/python3.9/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "Received RST_STREAM with error code 2"
	debug_error_string = "{"created":"@1660664399.127751000","description":"Error received from peer ipv4:142.250.0.95:443","file":"src/core/lib/surface/call.cc","file_line":967,"grpc_message":"Received RST_STREAM with error code 2","grpc_status":13}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/nick.bellamy/test/testing.py", line 4, in <module>
    response = client.get_connection_profile(request=request)
  File "/Users/nick.bellamy/test/datastream/lib/python3.9/site-packages/google/cloud/datastream_v1/services/datastream/client.py", line 742, in get_connection_profile
    response = rpc(
  File "/Users/nick.bellamy/test/datastream/lib/python3.9/site-packages/google/api_core/gapic_v1/method.py", line 154, in __call__
    return wrapped_func(*args, **kwargs)
  File "/Users/nick.bellamy/test/datastream/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 52, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InternalServerError: 500 Received RST_STREAM with error code 2
```

**Comment by vam-google** [comment]
Hi @nbellamy-paper,
Are you using any proxies? The grpc error code you are getting is `13` which is reserved for internal/serious problems. Basically something is not workign for you on fundamental level. 

This stackwoverflow [stackwoverflow response](https://stackoverflow.com/questions/57728710/how-to-fix-received-rst-stream-with-error-code-2-when-using-nginx-reverse-prox) response may help (unless you haven't seen it already).

In general, the issue seems to be on server/networking level which is beyond what client side library can control in a reasonable way. Also it is very suspicious that it fails on a basic example (which suggest that the isue is in environment or server side setup, not in how you use the library itself).

**Comment by nbellamy-paper** [comment]
Thanks for the response @vam-google 

No proxies in use, no. No VPN - just a direct internet connection. 

One of my colleagues (on a different laptop, in a different city & ISP) has the same problems with this library.

I can use other google python libs just fine.

```python
nick.bellamy ~  $ python3
Python 3.10.6 (main, Aug 11 2022, 13:36:31) [Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from google.cloud import bigquery
>>> for dataset in bigquery.Client().list_datasets():
...   print(dataset.dataset_id)
...
# redacted list of datasets - no error, output as expected
>>> from google.cloud import storage
>>> for bucket in storage.Client().list_buckets():
...   print(bucket.name)
...
# redacted list of buckets - no error, output as expected
>>> from google.cloud import datastream
>>> datastream.DatastreamClient().list_connection_profiles()
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 50, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/opt/homebrew/lib/python3.10/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "Received RST_STREAM with error code 2"
	debug_error_string = "{"created":"@1661956364.181316000","description":"Error received from peer ipv4:142.250.68.138:443","file":"src/core/lib/surface/call.cc","file_line":967,"grpc_message":"Received RST_STREAM with error code 2","grpc_status":13}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/lib/python3.10/site-packages/google/cloud/datastream_v1/services/datastream/client.py", line 640, in list_connection_profiles
    response = rpc(
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/gapic_v1/method.py", line 154, in __call__
    return wrapped_func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 52, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InternalServerError: 500 Received RST_STREAM with error code 2
>>>
```

What is the address `142.250.68.138` in `Error received from peer ipv4:142.250.68.138:443`?  I had assumed it was the GCP datastream API, and that I have no control over that endpoint. Is that something that I can control? 

I am happy to troubleshoot client/environment problems if you have further guidance, but everything else seems to be working fine. I don't have a web proxy on my laptop config, and I'm not aware of any reverse proxies that we control as described in the stackoverflow article.

**Comment by parthea** [comment]
Hi @nbellamy-paper, I recently came across the same issue. The value for `name` should be in the form `projects/<projectname>/locations/<location>/connectionProfiles/<profilename>` instead of `<profilename>`

As an example,
```
import google.cloud.datastream_v1 as datastream_v1
client = datastream_v1.DatastreamClient()
request = datastream_v1.GetConnectionProfileRequest(name="projects/someproject/locations/us-central1/connectionProfiles/mysql-bq-dev")
response = client.get_connection_profile(request=request)
```

I'll keep this issue open so we can track improving the sample code.

**Comment by nbellamy-paper** [comment]
I still receive the same error using that name format.  Still the same error on `list_connection_profiles()` as well.

```
$ python
Python 3.10.6 (main, Aug 11 2022, 13:36:31) [Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import google.cloud.datastream_v1 as datastream_v1
>>> client = datastream_v1.DatastreamClient()
>>> client.list_connection_profiles()
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 50, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/opt/homebrew/lib/python3.10/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "Received RST_STREAM with error code 2"
	debug_error_string = "{"created":"@1664832422.162304000","description":"Error received from peer ipv4:142.251.32.202:443","file":"src/core/lib/surface/call.cc","file_line":967,"grpc_message":"Received RST_STREAM with error code 2","grpc_status":13}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/lib/python3.10/site-packages/google/cloud/datastream_v1/services/datastream/client.py", line 640, in list_connection_profiles
    response = rpc(
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/gapic_v1/method.py", line 154, in __call__
    return wrapped_func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 52, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InternalServerError: 500 Received RST_STREAM with error code 2
>>> request = datastream_v1.GetConnectionProfileRequest(name="projects/redacted/locations/redacted/connectionProfiles/mysql-bq-dev")
>>> response = client.get_connection_profile(request=request)
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 50, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/opt/homebrew/lib/python3.10/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "Received RST_STREAM with error code 2"
	debug_error_string = "{"created":"@1664832504.684270000","description":"Error received from peer ipv4:142.251.32.202:443","file":"src/core/lib/surface/call.cc","file_line":967,"grpc_message":"Received RST_STREAM with error code 2","grpc_status":13}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/lib/python3.10/site-packages/google/cloud/datastream_v1/services/datastream/client.py", line 746, in get_connection_profile
    response = rpc(
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/gapic_v1/method.py", line 154, in __call__
    return wrapped_func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 52, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InternalServerError: 500 Received RST_STREAM with error code 2
>>>
```

**Comment by Rstar1998** [comment]
Even I am getting the same error : https://github.com/googleapis/python-datastream/issues/137
Even the REST API is not working for me. @nbellamy-paper did rest api work for you ?

**Comment by parthea** [comment]
Hi @nbellamy-paper,

I was able to re-create the error using the code provided.

```
>>> import google.cloud.datastream_v1 as datastream_v1
>>> client = datastream_v1.DatastreamClient()
>>> client.list_connection_profiles()
```

I modified the code to include the request parameter and I no longer receive the error. Please can you try the sample code below?
```

import google.cloud.datastream_v1 as datastream_v1
project_id = "your-project"
location = "us-central1"
client = datastream_v1.DatastreamClient()
request = datastream_v1.ListConnectionProfilesRequest(parent=f"projects/{project_id}/locations/{location}/connectionProfiles")
response = client.list_connection_profiles(request=request)
print(response)

```

**Comment by parthea** [comment]
Hi @Rstar1998,

It appears that this specific API provides a generic error code which is not helpful. Usually `Received RST_STREAM with error code 2` indicates that something is wrong with the request.
I've used the `gcloud` tool to troubleshoot the correct format required. 

For example, the following command may give a hint about the format needed for `parent`.
```
 gcloud datastream connection-profiles list --location=us-central1 --verbosity=debug
 ```
 
 Please feel free to open a new issue if you're still experiencing an issue.

**Comment by parthea** [comment]
I'm going to close this issue but please feel free to open a new issue if you're still experiencing a problem.

**Comment by nbellamy-paper** [comment]
I still receive the same error using the code you provided @parthea .

```
>>> import google.cloud.datastream_v1 as datastream_v1
>>> project_id = "redacted"
>>> location = "redacted"
>>> client = datastream_v1.DatastreamClient()
>>> request = datastream_v1.ListConnectionProfilesRequest(parent=f"projects/{project_id}/locations/{location}/connectionProfiles")
>>> response = client.list_connection_profiles(request=request)
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 50, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/opt/homebrew/lib/python3.10/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "Received RST_STREAM with error code 2"
	debug_error_string = "{"created":"@1665586914.027808000","description":"Error received from peer ipv4:142.251.46.138:443","file":"src/core/lib/surface/call.cc","file_line":967,"grpc_message":"Received RST_STREAM with error code 2","grpc_status":13}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/lib/python3.10/site-packages/google/cloud/datastream_v1/services/datastream/client.py", line 640, in list_connection_profiles
    response = rpc(
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/gapic_v1/method.py", line 154, in __call__
    return wrapped_func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 52, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InternalServerError: 500 Received RST_STREAM with error code 2
>>>
```

**Comment by parthea** [comment]
Hi @nbellamy-paper,

As I mentioned in https://github.com/googleapis/python-datastream/issues/123#issuecomment-1274830550, there is a known issue where the backend provides a generic error code `500` which is not helpful. Usually `Received RST_STREAM with error code 2` indicates that something is wrong with the request.

See the below example using curl, I get a valid response when I use location `us-central1` and I receive a 404 when I use location `us-east4`.

This succeeds
```
curl -X GET -H "Authorization: Bearer "$(gcloud auth application-default print-access-token) "https://datastream.googleapis.com/v1/projects/project-xxxx/locations/us-central1/streams/my-stream/objects"
```

This fails with 404 error
```
curl -X GET -H "Authorization: Bearer "$(gcloud auth application-default print-access-token) "https://datastream.googleapis.com/v1/projects/project-xxxx/locations/us-central1/streams/my-stream/objects"
```

With the python client, if I repeat the same steps I get a valid response with location `us-central1` but I receive 500 Received `RST_STREAM` with error code 2 when I use location `us-east4` which indicates that the server responds with 500 when it should respond with 404.

```
import google.cloud.datastream_v1 as datastream_v1
project_id ="partheniou-303520"
#location = "us-east4" This results in 500 Received RST_STREAM with error code 2
location="us-central1"
stream_name="my-stream"
client = datastream_v1.DatastreamClient()
request = datastream_v1.ListStreamObjectsRequest(parent=f"projects/{project_id}/locations/{location}/streams/{stream_name}")
response = client.list_stream_objects(request=request)
print(response)
```

This issue only happens when using the gRPC transport. We are actively developing a python client with REST transport, which could be used as a workaround however the work is still in progress. In the meantime, I've reported the issue to the API team. Googlers see b/246509350.

**Comment by parthea** [comment]
Closing this bug as obsolete. The issue was fixed b/246509350. Please open a new issue if the problem re-appears.
